rytilahti/python-miio
# Add basic support for Xiaomi Mi Air Purifier 3/3H

Add support for basic operations for new Xiaomi Air Purifier 3/3H (`zhimi-airpurifier-mb3` and `zhimi-airpurifier-ma4` (https://github.com/rytilahti/python-miio/issues/577)).

In order to support that, also implement `MiotDevice` class with basic support for MIoT protocol which the `AirPurifierMiot` extends to retrieve & set properties.

Mapping is based on https://github.com/rytilahti/python-miio/issues/543#issuecomment-544745035 - thanks, it helped a lot!

Note 
- I have not added any tests yet. I'd appreciate first round of a review before I do so.
- I have not tried the template/generative approach. Thought it would be better just to get it working. I can attempt to do that separately later.

Disclaimer: this is my first contribution to this project & I don't have Python experience.

ToDo:
- [x] extract Filter Type detection to a separate class

Manually tested operations:
- `info` (see output below)
- `status` (see output below)
- `off`
- `on`
- `set_buzzer <true|false>`
- `set_child_lock <true|false>`
- `set_fan_level <1-3>`
- `set_favorite_level <0-14>`
- `set_led <true|false>`
- `set_led_brightness <off|dim|bright>`
- `set_mode <auto|fan|favorite|silent>`

Example output of `info`:
```
$ python miio/cli.py airpurifiermiot --ip $AIR_IP --token $TOKEN info
Model: zhimi.airpurifier.mb3
Hardware version: esp32
Firmware version: 2.0.5
Network: {'localIp': '192.168.0.22', 'mask': '255.255.255.0', 'gw': '192.168.0.1'}
AP: {'ssid': 'coolwifi', 'bssid': '20:E5:2A:3C:BC:82', 'rssi': -59, 'primary': 8}
```

Example output of `status`:
```
$ python miio/cli.py airpurifiermiot --ip $AIR_IP --token $TOKEN status
Power: on
AQI: 3 μg/m³
Average AQI: 6 μg/m³
Humidity: 58 %
Temperature: 25.799999 °C
Mode: OperationMode.Favorite
LED: True
LED brightness: LedBrightness.Dim
Buzzer: False
Child lock: False
Favorite level: 14
Filter life remaining: 98 %
Filter hours used: 59
Use time: 212700 s
Purify volume: 8606 m³
Motor speed: 2148 rpm
Filter RFID product id: 0:0:31:31
Filter RFID tag: 80:66:6a:da:38:bd:4
Filter type: FilterType.Regular
```

Closes #577
